File: src/Ccmc.Core/Services/ProjectNameValidator.cs
namespace Ccmc.Core.Services;

/// <summary>Validates a new project name against a destination root.</summary>
public static class ProjectNameValidator
{
    private static readonly char[] InvalidChars = ['<', '>', ':', '"', '/', '\\', '|', '?', '*'];

    /// <summary>Returns a user-facing error message, or null when the name is valid.</summary>
    public static string? GetError(string name, string root)
    {
        if (string.IsNullOrWhiteSpace(name))
            return "Project name cannot be empty.";
        if (name.IndexOfAny(InvalidChars) >= 0)
            return "Project name contains invalid characters: < > : \" / \\ | ? *";
        var target = Path.Combine(root, name.Trim());
        if (Directory.Exists(target) || File.Exists(target))
            return $"A folder named '{name.Trim()}' already exists in {root}.";
        return null;
    }

    public static string CreateProjectFolder(string root, string name)
    {
        if (!Directory.Exists(root))
            throw new DirectoryNotFoundException($"Root folder does not exist: {root}");
        var path = Path.Combine(root, name.Trim());
        Directory.CreateDirectory(path);
        return path;
    }
}

File: electron/src/core/projects/projectNameValidator.ts
/** Validates a new project name. Pure logic — filesystem checks are injected by caller. */

const INVALID_CHARS = ['<', '>', ':', '"', '/', '\\', '|', '?', '*']

/**
 * Returns a user-facing error message, or null when the name is valid.
 * @param name The candidate project name.
 * @param existsCheck Called with the trimmed name; returns true if a file/folder already exists.
 * @param root Optional root path for inclusion in the "already exists" message.
 */
function getError(
  name: string,
  existsCheck: (trimmedName: string) => boolean,
  root?: string
): string | null {
  if (!name || !name.trim()) {
    return 'Project name cannot be empty.'
  }

  for (const ch of INVALID_CHARS) {
    if (name.includes(ch)) {
      return 'Project name contains invalid characters: < > : " / \\ | ? *'
    }
  }

  const trimmed = name.trim()
  if (existsCheck(trimmed)) {
    const location = root ? ` in ${root}` : ''
    return `A folder named '${trimmed}' already exists${location}.`
  }

  return null
}

export const projectNameValidator = { getError, INVALID_CHARS }
